state/protocol/protocol_state/state/mutator.go

```diff
@@ -100,17 +92,33 @@ func newStateMutator(
 //     of the calling code (specifically `FollowerState`).
 //
 // updated protocol state entry, state ID and a flag indicating if there were any changes.
-func (m *stateMutator) Build() (hasChanges bool, updatedState *flow.ProtocolStateEntry, stateID flow.Identifier, dbUpdates []transaction.DeferredDBUpdate) {
-	updatedState, stateID, hasChanges = m.epochHappyPathStateMachine.Build()
-	dbUpdates = m.pendingDbUpdates
-	return
+func (m *stateMutator) Build() (flow.Identifier, []transaction.DeferredDBUpdate, error) {
+	var dbUpdates []transaction.DeferredDBUpdate
+	for _, stateMachine := range m.orthoKVStoreMachines {
+		dbUpdates = append(dbUpdates, stateMachine.Build()...)
+	}
+	stateID := m.kvMutator.ID()
+	version, data, err := m.kvMutator.VersionedEncode()
+	if err != nil {
+		return flow.ZeroID, nil, fmt.Errorf("could not encode protocol state: %w", err)
+	}
```

Comment:
😭 I think I understand now why you implemented it that way:
* not only do we want to index the KV-Store by the ID of the resulting block
* but we also want that individual sub-state machines can maintain their own `ByBlock` indices. That makes sense and is a very strong feature. 

For this purpose, I think `transaction.DeferredDBUpdate` is too simple:
* `transaction.DeferredDBUpdate` assumes that all information is already available at the time we construct the pending DB updates.
* However, this does not apply to the ID of the resulting block. 

I think we want something more powerful. A construct where we can say: "Once we are done with the construction and have the final block ID, _then_ we want to apply these updates and indexing operations"

I would suggest to structure this as a _wrapper_ around `transaction.DeferredDBUpdate`